Add RM_RdbLoad() and RM_RdbSave()

tezc opened the pull request:
Add `RM_RdbLoad()` and `RM_RdbSave()` to load/save RDB files from the module API. 

In our use case, we have our clustering implementation as a module. As part of this implementation, the module needs to trigger RDB save operation at specific points. Also, this module delivers RDB files to other nodes (not using Redis' replication). When a node receives an RDB file, it should be able to load the RDB. Currently, there is no module API to save/load RDB files. 


This PR adds four new APIs:
```c
RedisModuleRdbStream *RM_RdbStreamCreateFromFile(const char *filename);
void RM_RdbStreamFree(RedisModuleRdbStream *stream);

int RM_RdbLoad(RedisModuleRdbStream *stream, int flags);
int RM_RdbSave(RedisModuleRdbStream *stream, int flags);
```

The first step is to create a `RedisModuleRdbStream` object. This PR provides a function to create RedisModuleRdbStream from the filename. (You can load/save RDB with the filename). In the future, this API can be extended if needed: 
e.g., `RM_RdbStreamCreateFromFd()`, `RM_RdbStreamCreateFromSocket()` to save/load RDB from an `fd` or a `socket`. 


Usage:
```c
/* Save RDB */
RedisModuleRdbStream *stream = RedisModule_RdbStreamCreateFromFile("exp.rdb");
RedisModule_RdbSave(stream, 0);
RedisModule_RdbStreamFree(stream);

/* Load RDB */
RedisModuleRdbStream *stream = RedisModule_RdbStreamCreateFromFile("exp.rdb");
RedisModule_RdbLoad(stream, 0);
RedisModule_RdbStreamFree(stream);
```

tezc committed:
RM_RdbLoad(), RM_RdbSave()

oranagra commented on src/module.c:
```diff
@@ -12210,6 +12210,76 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
+        return REDISMODULE_ERR;
+    }
+
+    if (server.child_type == CHILD_TYPE_RDB) {
```
i don't think we care, as long as they don't save to the same file.
note that if this API is called from fork, we can't have two forks at the same time anyway.

tezc replied:
> i don't think we care, as long as they don't save to the same file.

Ok, what should we do if it is the same file? I added a check for it. If it is the same file, returns EINPROGRESS.

oranagra replied:
What about taking a file descriptor instead of a name?

oranagra commented on src/module.c:
```diff
@@ -12210,6 +12210,76 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
+        return REDISMODULE_ERR;
+    }
+
+    server.stat_rdb_saves++;
```
i don't think we wanna update stats, we don't know where this file is saved to, from my perspective, it's not part of redis's persistence mechanism.

tezc replied:
ok, removed

oranagra commented on src/module.c:
```diff
@@ -12210,6 +12210,76 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
+    if (server.current_client)
+        unprotectClient(server.current_client);
+
+    /* TODO:
+     * - rename(filename, server.rdb_filename) ?;
+     * - Start AOF?
```
yes, we should probably stop and re-start AOF.
regarding the rename, i think that's the module's responsibility to hand over a temp file name and do any rename or even an fsync..
while we're on that subject, maybe we should take in a file descriptor and not a file name?

tezc replied:
Ok, leaving rename() to the module.

oranagra commented on src/module.c:
```diff
@@ -12210,6 +12210,76 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
+        return REDISMODULE_ERR;
+    }
+
+    /* TODO:
+     * - Call fsync for filename
+     * - Stop AOF?
+     * - Kill RDB fork?
+     */
```
yes, let's stop the aof (restart later), and let's kill the rdb fork (storing outdated data, and creating COW mess).

why would we want to fsync before storing anything?

tezc replied:
done

oranagra reviewed (commented):
i think it's a good idea to also document these and their purpose (that's a good way to validate the API design).

i think both should probably take a file descriptor rather than a file name, and let the module do the rest (more flexible)

obviously add tests and a detailed top comment when ready..

tezc committed:
v1

tezc committed:
v2

tezc committed:
test fix

tezc committed:
comment

tezc committed:
better log line in test

tezc commented:
@oranagra Updated the PR. Added tests and docs. Please take a look. 

Regarding calling `fsync()` in `RM_RdbLoad()`, I thought same as if like we receive RDB from network. I added `fsync()` to make it durable (not trusting the module) but I guess it is better to leave it to the module. So, no `fsync()` anywhere. 

Regarding using `filename` vs `fd` in the API. Isn't filename better if we want to do some checks? Normally, we can open the file if we have the filename but we cannot learn filename from fd. For example, please see `RM_RdbSave()`, I added a check if we are saving to the same file with the bgsave. I'm not sure if we want that specific check but if we are going to have `fd` in the API, we won't be able to add these kind of checks (now or in future). Somehow I feel like filename is more flexible but maybe that's just me :) Please comment what you like better, I'll change the PR.

oranagra commented:
The advantage of an fd is that it can be a pipe or socket. I don't think we need that filename conflict check. We can document any concerns instead

tezc commented:
Ok, I'll change filename to fd.

oranagra commented on src/module.c:
```diff
@@ -12210,47 +12210,64 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
 
-    if (server.current_client)
-        protectClient(server.current_client);
+    /* Kill existing RDB fork as it is saving outdated data. */
```
we can also mention we do it to avoid COW hell

tezc replied:
done

oranagra commented on tests/unit/moduleapi/rdbloadsave.tcl:
```diff
@@ -0,0 +1,151 @@
+        # cause a problem.
+        # e.g. Redis won't try to process next message of the current client
+        # while it is in the command callback.
+        r write [format_command test.rdbload blabla.rdb]
+        r flush
+        r write [format_command get x]
+        r write [format_command dbsize]
+        r flush
```
if you really want them to be processed in the same event loop better write them with a single `write` call, or even better, make sure to SIGSTOP and SIGCONT before and after writing.

tezc replied:
I just want it to try to process next message while it is inside `rdbLoad()`. 

I tried SIGSTOP/SIGCONT, looks like it won't trigger the issue I saw (Currently, with the above test, it triggers an assert without `protectClient()` inside `RM_RdbLoad()`). 

Probably, the reason is if it reads all messages to the client buffer, then, when `RM_RdbLoad()` goes back to the event loop, there is no new event for the client. So, it won't trigger the assert.

tezc commented:
Currently, API is 

```c
int RM_RdbSave(const char *filename, int flags);
int RM_RdbLoad(const char *filename, int flags);
```

Oran thinks accepting `fd` would be more flexible. Module can pass a pipe or socket as well: 

```c
int RM_RdbSave(int fd, int flags);
int RM_RdbLoad(int fd, int flags);
```

To do that, feels like using `rioInitWithFd()` in Redis is the best approach. Things to do in Redis: 
- `rioFdIO` is missing read function. We need to implement it. 
- `rioFdWrite()` is missing incremental fsync. Eventually, we will need it for redisraft for the same purpose (to avoid fsync stall)
- If we change to `fd` in the API, we cannot assume it is a file and add incremental fsync blindly. 
       To implement it, module should tell us whether it's a file or not. (needs to pass a flag)

So, right now question is which API is better?
- If we go with `fd`, it will be harder to detect if it is a pipe, file or socket. (module should pass a flag).
- If we go with `filename`, module cannot pass socket or pipe. 

@yossigo ping. 

----
**Edit**: Maybe there is a way to detect if an `fd` is a file or not:

https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/stat.h.html

```
S_ISBLK(m)
   Test for a block special file.
S_ISCHR(m)
    Test for a character special file.
S_ISDIR(m)
    Test for a directory.
S_ISFIFO(m)
   Test for a pipe or FIFO special file.
S_ISREG(m)
   Test for a regular file.
S_ISLNK(m)
   Test for a symbolic link.
S_ISSOCK(m)
   Test for a socket.
```

I didn't use this API before, not sure about any possible complexities.

---

yossigo commented:
@tezc I agree with @oranagra that it makes sense to think beyond simply reading or writing a file, but a file descriptor might not offer enough abstraction. For example, it works if it's a TCP socket - but realistically, it may need to be TLS, and then a file descriptor is no longer useful.

So, eventually, we may want to conceptually expose `rio`, so a module could implement the different callbacks. But that's overkill for what we need now. A compromise that is easy now but should be extensible in the future can be:

* Introduce an opaque `RDBStream` type, and accept it instead of a file descriptor/name.
* Use a `RM_CreateRDBStreamFile(filename)` that creates the above structure and populates it with the filename.
* In the future, we can have `RM_CreateRDBStreamFd` to implement the fd option or `RM_CreateRDBStreamCustom` that lets the module provide full rio-like callbacks.

Does this make sense?

tezc commented:
Ok, makes sense. I'll update the PR.

tezc committed:
Use RedisModuleRdbStream

tezc committed:
Style

tezc commented on src/module.c:
```diff
@@ -12210,6 +12210,141 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
+    rdbSaveInfo rsi, *rsiptr;
+    rsiptr = rdbPopulateSaveInfo(&rsi);
+
+    if (rdbSave(SLAVE_REQ_NONE,stream->data.filename,rsiptr) != C_OK) {
```
https://github.com/redis/redis/blob/ad72cb77975a54bc1c4b1370835c7803b1ce9439/src/rdb.c#L1502

```c
    server.dirty = 0;
    server.lastsave = time(NULL);
    server.lastbgsave_status = C_OK;
    stopSaving(1);
```
I have a question here. I call `rdbSave()` here directly and it sets above variables internally. 
If we don't see this new `RM_RdbSave()` as part of Redis persistence mechanism, probably we should not set those above variables? Should we even call callbacks through startSaving()/stopSaving()?

@oranagra wdyt?

oranagra replied:
you're right. since we don't save the rdb file to the designated persistence location, then we should not mess up the status metrics (dirty, lastsave, etc).
but we should still call startSaving, to set the current state.
i.e. modules should be aware that we're going into a state of generating persistence (they may need to do some prep before being called to serialize their data)

oranagra replied:
so i see you moved startSaving and stopSaving to the outer function, so that concern is resolved (no implications on the existing behavior, right?)
i would be ok with putting them inside the inner function too, but no objections to what you've done

tezc committed:
Add rdbSaveToFile()

oranagra commented on src/rdb.c:
```diff
@@ -1436,6 +1436,48 @@ int rdbSaveRioWithEOFMark(int req, rio *rdb, int *error, rdbSaveInfo *rsi) {
     return C_ERR;
 }
 
+/* Save DB to the file. Similar to rdbSave() but this function won't use a
+ * temporary file and won't update the metrics. */
+int rdbSaveToFile(const char *filename) {
```
maybe there's a cleaner way to avoid code duplication by changing rdbSave to use rdbSaveToFile?
i.e. it generates the temp file name, and later does the rename and update metrics..

tezc replied:
I'm having a bit hard time to reuse this code. Currently, `rdbSave()` updates the metrics and calls `stopSaving()`. If we call `rdbSaveToFile()` first, it will call `stopSaving()` before updating the metrics. I thought this might be a breaking change (module callback will be called before updating the metrics), that's why I went with separate functions. If you think this is not important, I can make the change.

oranagra replied:
you mean `dirty`, `lastsave`, and `lastbgsave_status`?
i don't see a problem with the last two, modules do get an indication of success in the callback.
maybe the only risk here is that is set reset the dirty counter after the callback, then it'll undo any `dirty++` that's done inside the callback, i.e. if the module would do call RM_Call from that callback.
bottom line, i think modules are not suppose to do any such things from this callback, so re-ordering this seems ok to me.

tezc replied:
Yeah, I was worried about things like if module wants to read `lastsave` timestamp in the callback (via `RM_Call("info")` maybe?). If we don't care about these, then I'll update the PR.

tezc replied:
@oranagra added a commit for this. Please take a look.

oranagra replied:
so i see you moved startSaving and stopSaving to the outer function, so that concern is resolved (no implications on the existing behavior, right?)
i would be ok with putting them inside the inner function too, but no objections to what you've done

tezc replied:
Yes, I moved them to the outer function. Maybe a tiny implication, previously, we were calling `startSaving()` after `fopen()`. Now, `startSaving()` will be called after `fopen()`. 

If we put `startSaving() / stopSaving()` into the internal function, for `rdbSave()`, we would call `stopSaving(1)` (with success code) before doing rename + fsync dir. So, right now, it will only call `stopSaving()` after rename + fsync.

oranagra replied:
ok. LGTM.

tezc committed:
Reuse rdbSaveToFile()

tezc committed:
Minimize diff a bit

tezc committed:
Remove unused vars

oranagra commented on src/rdb.c:
```diff
@@ -1475,7 +1470,39 @@ int rdbSave(int req, char *filename, rdbSaveInfo *rsi) {
+    startSaving(RDBFLAGS_NONE);
+    snprintf(tmpfile,256,"temp-%d.rdb", (int) getpid());
+
+    if (rdbSaveInternal(req,tmpfile,rsi) != C_OK) goto werr;
```
if rdbSaveInternal failed, we don't need to attempt to unlink the tempfile in the error handling.
same for when fsyncFileDir fails.
in both cases it'll just fail to unlink, so maybe we don't care since we're gonna ignore the error anyway.
WDYT?

tezc replied:
You mean we can leave that file there? 

Just to clean up temporary file, maybe we can separate parts. 
- We don't call unlink() if `rdbSaveInternal()` fails. (As it will call unlink() internally).
- We call `unlink()` if rename() or fsyncFileDir() fails.

oranagra replied:
i mean that there's no temporary files.
if rdbSaveInternal failed, it already deletes it inside the function.
and if fsyncFileDir it fails after the rename, so the temporary file is already gone.

i think we should keep the unlink inside rdbSaveInternal, and either add a boolean flag to indicate when the unlinking is needed, or keep the unlink and add a comment that it may fail because the file is already deleted / renamed.

i'm not sure i care much about what happens i fsyncFileDir fails (as long as it prints an error).
not even sure we should report the error to the caller, but if we do and the file remains that's fine i think.

tezc replied:
Okay, now I see your point. 

Added a commit. Maybe we just call unlink() on rename failure only. 
If fsyncFileDir() fails, there is nothing to do anyway. We don't unlink().

tezc committed:
Precise unlink()
